internal/proxy/proxy.go

```diff
@@ -289,6 +311,15 @@ func (c *Client) serveSocketMount(ctx context.Context, s *socketMount) error {
 		go func() {
 			c.cmd.Printf("[%s] accepted connection from %s\n", s.inst, cConn.RemoteAddr())
 
+			count := atomic.AddUint32(&c.connCount, 1)
+			defer atomic.AddUint32(&c.connCount, ^uint32(0))
+
+			if c.maxConns > 0 && count > c.maxConns {
+				c.cmd.Printf("max connections (%v) exceeded, refusing new connection\n", c.maxConns)
+				cConn.Close()
```

Comment:
The proxy is closing connections. This gives the client no indication that the server is too busy. I would expect the proxy to behave in a way that provides better hints to the client that it is not accepting connections. 

Two options I would consider better behaved:

- refuse to accept the connection by closing the listening socket. Clients will get "Connection Refused" or "EOF" from their net.Dial() call. This is fail-fast. 
- allow the inbound connections to queue up in the inbound server connection queue by only calling `s.listener.Accept()` when connections < maxConnections. On the client side, net.Dial() will hang until the server accepts the connection or client-side tcp connect timeout is exceeded.  This is fail-slow.

I wrote a simple example executable demonstrating how to implement both server-side options in go: 
[main.go.txt](https://github.com/GoogleCloudPlatform/cloudsql-proxy/files/8962108/main.go.txt)

Reply:
So this is a reasonable thought, but the way it's currently implemented does match what's being done in v1: https://github.com/GoogleCloudPlatform/cloudsql-proxy/blob/957d1609ad96bfed77b3744f1c11a762010bc06e/proxy/proxy/client.go#L177 (that doesn't mean we shouldn't discuss if this is the behavior we want though).

For the current behavior, clients do get a "server closed the connection" and if they check the proxy logs, they'll see why the connection was closed. 

For the first alternative, I think that closing/opening the socket will introduce a lot of overhead (it's essentially a syscall). We'd have to reopen/close (possible several) sockets all at once any time a connection opens or closes. It's also potentially harder to debug, as there won't be a way for the proxy to tell when this caused a client side connection to fail. 

For the second alternative, this seems similar to how [netutil.LimitListener](https://cs.opensource.google/go/x/net/+/263ec571:netutil/listen.go;l=16) works. The one caveat is that this for just 1 socket, where we need to share across all of them. This probably has the same drawbacks -- it'll end up being hard to debug and might cause issues.

So my hot take is the current behavior seems preferable. thoughts?

Reply:
Let's rule out Option 1. I agree with your reasoning about syscalls.

Going ahead with an implementation matching V1 is viable.  Changing to an Option 2 implementation that respects a global active connection limit is also viable and for me, closer to what I expect well-behaved proxy to do in theory. 

But more important than my opinion, what has happened in practice? In our experience with the V1 proxy, has the limit and connection behavior lead to support calls with confused customers or do they not really notice? Maybe V1 behavior is good enough for the real world.

Reply:
> Maybe V1 behavior is good enough for the real world.

This is my view. Granted, we don't have a lot of data on how frequently these flags are used. So we could choose:

1. to forgo implementing this flag until someone asks for it, or
2. match v1 behavior.

The guiding light here is that we want the migration to v2 to be as friction-less as possible, even if that means having flags that are no used heavily.